library/ssl_tls.c

```diff
@@ -2871,21 +2878,72 @@ int mbedtls_ssl_flight_transmit( mbedtls_ssl_context *ssl )
             ssl_swap_epochs( ssl );
         }
 
-        memcpy( ssl->out_msg, cur->p, cur->len );
-        ssl->out_msglen = cur->len;
-        ssl->out_msgtype = cur->type;
+        /* CCS is copied as is, while HS messages may need fragmentation */
+        if( cur->type == MBEDTLS_SSL_MSG_CHANGE_CIPHER_SPEC )
+        {
+            memcpy( ssl->out_msg, cur->p, cur->len );
+            ssl->out_msglen = cur->len;
+            ssl->out_msgtype = cur->type;
+
+            /* Update position inside current message */
+            ssl->handshake->cur_msg_p += cur->len;
+        }
+        else
+        {
+            const unsigned char * const p = ssl->handshake->cur_msg_p;
+            const size_t hs_len = cur->len - 12;
+            const size_t frag_off = p - ( cur->p + 12 );
+            const size_t rem_len = hs_len - frag_off;
+            const size_t frag_len = rem_len > HS_LIMIT ? HS_LIMIT : rem_len;
+
+            /* Messages are stored with handshake headers as if not fragmented,
+             * copy beginning of headers then fill fragmentation fields.
+             * Handshake headers: type(1) len(3) seq(2) f_off(3) f_len(3) */
+            memcpy( ssl->out_msg, cur->p, 6 );
+
+            ssl->out_msg[6] = ( ( frag_off >> 16 ) & 0xff );
+            ssl->out_msg[7] = ( ( frag_off >>  8 ) & 0xff );
+            ssl->out_msg[8] = ( ( frag_off       ) & 0xff );
 
-        ssl->handshake->cur_msg = cur->next;
+            ssl->out_msg[ 9] = ( ( frag_len >> 16 ) & 0xff );
+            ssl->out_msg[10] = ( ( frag_len >>  8 ) & 0xff );
+            ssl->out_msg[11] = ( ( frag_len       ) & 0xff );
 
-        MBEDTLS_SSL_DEBUG_BUF( 3, "handshake header", ssl->out_msg, 12 );
+            MBEDTLS_SSL_DEBUG_BUF( 3, "handshake header", ssl->out_msg, 12 );
+
+            /* Copy the handshame message content and set records fields */
+            memcpy( ssl->out_msg + 12, p, frag_len );
+            ssl->out_msglen = frag_len + 12;
+            ssl->out_msgtype = cur->type;
+
+            /* Update position inside current message */
+            ssl->handshake->cur_msg_p += frag_len;
+        }
+
+        /* If done with the current message move to the next one if any */
+        if( ssl->handshake->cur_msg_p >= cur->p + cur->len )
+        {
+            if( cur->next != NULL )
+            {
+                ssl->handshake->cur_msg = cur->next;
```

Comment:
If `mbedtls_ssl_write_record()` returns `MBEDTLS_ERR_SSL_WANT_WRITE`, then we'll return to the application and next time the application calls a function from the SSL module, the first this this function will do is call `mbedtls_ssl_flush_output()` in order to finish sending, well before we reach the current code path again. Now the pointers in question are only used in the current function, not in `mbedtls_ssl_flush_output()`, so they can have any value we want.

However, thanks for pointing that out, because I think I had actually forgotten to exercise that in the tests, so I've now added two tests with `nbio=2` in order to make sure the code behaves correctly in this case.

Reply:
@mpg Thanks for the clarification. My first thought when I saw it that these pointers could have been made local, and only then I considered they may also be kept as the SSL state. Does it mean that they can be moved to the local scope here?

Reply:
No, I don't think they could be made local, they need to be remembered if this function is exited and re-entered later. The simple case would be if `mbedtls_ssl_write_record()` always returned zero: in that case, we would always go to the end of the loop in one call, and the loop variables could indeed be local to the function in which the loop resides.

Unfortunately, we also need to support non-blocking I/O, where `mbedtls_ssl_write_record()` might return a temporary error at any time: `MBEDTLS_ERR_SSL_WANT_WRITE`. Here our model mandates that we return control to the application, that is we need to break the loop, return from the current function, its caller, all the way up to the first function in the call stack that belongs to the application. When any function from the SSL module is called later, it needs to have enough state saved in the SSL context in order to (1) try the underlying `f_send()` call that failed temporarily, and if it succeeds (2) call again the function that was being called (eg retransmitting a flight), which will itself do the same, until we reach `mbedtls_flight_transmit()` that then needs to (3) continue the loop where it was as if nothing happened.

That's why the loop variables (in addition to some other information, such as DTLS (re)transmission state) need to be persistent (saved into the context) rather than local variables. I guess one could call this a limited form of [continuation](https://en.wikipedia.org/wiki/Continuation) embedded in the SSL context.

It took me some time to get used to it when I first started to work on the project, and even now, I'm never really confident I got it right until I passed some relevant tests with `nbio=2` (special I/O callbacks which I wrote specifically to test this). Fortunately it looks like in that case I got it right, and only forgot to add tests for it later, so thanks for the reminder!